text/0000-futures.md

```diff
@@ -0,0 +1,617 @@
+`Future` has the need to get polled again, it can use the `wake` method of the
+waker in order to inform the executor that the task which owns the `Future`
+should get scheduled and executed again.
+
+The RFC defines a concrete `Waker` type with which implementors of `Futures`
+and asynchronous functions will interact. This type defines a `wake(&self)`
+method which is used to schedule the task that is associated to the `Waker`
+to be polled again.
+
+The mechanism through which tasks get scheduled again depends on the executor
+which is driving the task.
+Possible ways of waking up a an executor include:
+- If the executor is blocked on a condition variable, the condition variable
+  needs to get notified.
+- If the executor is blocked on a system call like `select`, it might need
+  to get woken up by a syscall like `write` to a pipe.
+- If the executor's thread is parked, the wakeup call needs to unpark it.
+
+To allow executors to implement custom wakeup behavior, the `Waker` type
+contains a type called `RawWaker`, which consists of a  pointer
+to a custom wakeable object and a reference to a virtual function
+pointer table (vtable) which provides functions to `clone`, `wake`, and
+`drop` the underlying wakeable object.
+
+This mechanism is chosen in favor of trait objects since it allows for more
+flexible memory management schemes. `RawWaker` can be implemented purely in
+terms of global functions and state, on top of reference counted objects, or
+in other ways. This strategy also makes it easier to provide different vtable
+functions that will perform different behaviors despite referencing the same
+underlying wakeable object type.
+
+The relation between those `Waker` types is outlined in the following definitions:
+
+```rust
+/// A `RawWaker` allows the implementor of a task executor to create a `Waker`
+/// which provides customized wakeup behavior.
+///
+/// It consists of a data pointer and a virtual function pointer table (vtable) that
+/// customizes the behavior of the `RawWaker`.
+#[derive(PartialEq)]
+pub struct RawWaker {
+    /// A data pointer, which can be used to store arbitrary data as required
+    /// by the executor. This could be e.g. a type-erased pointer to an `Arc`
+    /// that is associated with the task.
+    /// The value of this field gets passed to all functions that are part of
+    /// the vtable as first parameter.
+    pub data: *const (),
+    /// Virtual function pointer table that customizes the behavior of this waker.
+    pub vtable: &'static RawWakerVTable,
+}
+
+/// A virtual function pointer table (vtable) that specifies the behavior
+/// of a `RawWaker`.
+///
+/// The pointer passed to all functions inside the vtable is the `data` pointer
+/// from the enclosing `RawWaker` object.
+#[derive(PartialEq, Copy, Clone)]
+pub struct RawWakerVTable {
+    /// This function will be called when the `RawWaker` gets cloned, e.g. when
+    /// the `Waker` in which the `RawWaker` is stored gets cloned.
```

Comment:
This feels a unspecified; Under what circumstances is it undefined behavior to call `(table.clone)(ptr)`, `(table.wake)(ptr)`, and `(table.drop_fn)(ptr)`? That should be added to the documentation.